library/std/src/io/buffered/bufreader.rs

```diff
@@ -171,7 +171,7 @@ impl<R> BufReader<R> {
     #[stable(feature = "bufreader_buffer", since = "1.37.0")]
     pub fn buffer(&self) -> &[u8] {
         // SAFETY: self.cap is always <= self.init, so self.buf[self.pos..self.cap] is always init
-        unsafe { MaybeUninit::slice_assume_init_ref(&self.buf[self.pos..self.cap]) }
+        unsafe { MaybeUninit::slice_assume_init_ref(&self.buf.get_unchecked(self.pos..self.cap)) }
```

Comment:
I feel like this safety comment needs to be improved... but with what?

Reply:
We sort of already had the problem of potentially reading uninit memory with this before, but at least it was guaranteed to be *this* buffer. With this change any bug to setting pos/cap correctly will cause us to read potentially entirely different memory, which seems unfortunate. While BufReader's logic is probably OK, I'm not confident enough in our current tests or future tests that we'd catch any newly-introduced bugs before they hit someone else down the line.

This is basically the same point as your safety comment question -- it's really hard to be sure about correctness across a fairly large module for maintaining these values appropriately.

I think what might help here is moving the BufReader definition into a submodule, and then having a much more limited API surface area on that inner type that we can more clearly state preserves the properties we need here. It's probably not even that large a refactor, or at least I'd hope we're largely already making using of the functions that make up the 'core' interface in the helper functions we define. Maybe just fill_buf, buffer, and consume?

Reply:
I gave this a shot. The separation is not as clean as I would have liked, and this defeats one of the optimization opportunities I identified originally, which was omitting the check in `consume` when we are certain we are consuming a number of bytes in-bounds. Would it make sense to have a `consume_unchecked`? That could bring back this optimization, though I'm not sure if the unsafety is worth it, being that I really feel like this is hacking around some LLVM imprecision.